File: uis/client/web2/src/services/biometric_service.ts
import { AppConfig } from '../config/app_config';

export class BiometricService {
  // Check if biometric authentication is available (WebAuthn)
  async isBiometricAvailable(): Promise<boolean> {
    try {
      return !!(window.PublicKeyCredential && navigator.credentials);
    } catch {
      return false;
    }
  }

  // Get available biometric types (WebAuthn doesn't expose types reliably)
  async getAvailableBiometrics(): Promise<string[]> {
    const available: string[] = [];
    if (await this.isBiometricAvailable()) {
      // Browsers don't expose exact biometric type; assume generic
      available.push('Biometric');
    }
    return available;
  }

  // Check if device supports biometrics
  async isDeviceSupported(): Promise<boolean> {
    return this.isBiometricAvailable();
  }

  // Authenticate with biometrics using WebAuthn (simplified)
  async authenticate(_localizedReason = 'Please authenticate to access your account'): Promise<boolean> {
    if (!(await this.isBiometricAvailable())) return false;

    try {
      // Dummy credential request (real implementation would use a server challenge)
      const credential = await navigator.credentials.get({
        publicKey: {
          challenge: new Uint8Array(32),
          timeout: 60000,
          allowCredentials: [],
          userVerification: 'required',
        },
      });
      return !!credential;
    } catch {
      return false;
    }
  }

  // Check if biometric is enabled for this app
  isEnabled = async (): Promise<boolean> => {
    return localStorage.getItem(AppConfig.biometricEnabledKey) === 'true';
  };

  // Enable biometric
  enable = async (): Promise<void> => {
    localStorage.setItem(AppConfig.biometricEnabledKey, 'true');
  };

  // Disable biometric
  disable = async (): Promise<void> => {
    localStorage.removeItem(AppConfig.biometricEnabledKey);
  };

  isBiometricEnabled(): boolean {
    return localStorage.getItem(AppConfig.biometricEnabledKey) === 'true';
  }

  // Enable biometric authentication
  enableBiometric(): void {
    localStorage.setItem(AppConfig.biometricEnabledKey, 'true');
  }

  // Disable biometric authentication
  disableBiometric(): void {
    localStorage.setItem(AppConfig.biometricEnabledKey, 'false');
  }

  // Get biometric type name
  getBiometricTypeName(biometrics: string[]): string {
    if (biometrics.includes('Face ID')) return 'Face ID';
    if (biometrics.includes('Fingerprint')) return 'Fingerprint';
    if (biometrics.includes('Iris')) return 'Iris';
    if (biometrics.includes('Biometric')) return 'Biometric';
    return 'Biometric';
  }
}

export const biometricService = new BiometricService();

File: uis/client/mobile_app/lib/services/biometric_service.dart
import 'package:local_auth/local_auth.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import '../config/app_config.dart';

class BiometricService {
  final LocalAuthentication _localAuth = LocalAuthentication();
  final FlutterSecureStorage _storage = const FlutterSecureStorage(
    aOptions: AndroidOptions(
      encryptedSharedPreferences: true,
      resetOnError: true,
    ),
  );

  // Check if biometric authentication is available
  Future<bool> isBiometricAvailable() async {
    try {
      return await _localAuth.canCheckBiometrics;
    } catch (e) {
      return false;
    }
  }

  // Get available biometric types
  Future<List<BiometricType>> getAvailableBiometrics() async {
    try {
      return await _localAuth.getAvailableBiometrics();
    } catch (e) {
      return [];
    }
  }

  // Check if device supports biometrics
  Future<bool> isDeviceSupported() async {
    try {
      return await _localAuth.isDeviceSupported();
    } catch (e) {
      return false;
    }
  }

  // Authenticate with biometrics
  Future<bool> authenticate({
    String localizedReason = 'Please authenticate to access your account',
    bool useErrorDialogs = true,
    bool stickyAuth = true,
  }) async {
    try {
      final isAvailable = await isBiometricAvailable();
      if (!isAvailable) return false;

      return await _localAuth.authenticate(
        localizedReason: localizedReason,
        options: AuthenticationOptions(
          stickyAuth: stickyAuth,
          biometricOnly: true,
          useErrorDialogs: useErrorDialogs,
        ),
      );
    } catch (e) {
      return false;
    }
  }

  // Check if biometric is enabled for this app
  Future<bool> isBiometricEnabled() async {
    final enabled = await _storage.read(key: AppConfig.biometricEnabledKey);
    return enabled == 'true';
  }

  // Enable biometric authentication
  Future<void> enableBiometric() async {
    await _storage.write(key: AppConfig.biometricEnabledKey, value: 'true');
  }

  // Disable biometric authentication
  Future<void> disableBiometric() async {
    await _storage.write(key: AppConfig.biometricEnabledKey, value: 'false');
  }

  // Get biometric type name
  String getBiometricTypeName(List<BiometricType> biometrics) {
    if (biometrics.contains(BiometricType.face)) {
      return 'Face ID';
    } else if (biometrics.contains(BiometricType.fingerprint)) {
      return 'Fingerprint';
    } else if (biometrics.contains(BiometricType.iris)) {
      return 'Iris';
    } else if (biometrics.contains(BiometricType.strong)) {
      return 'Biometric';
    } else if (biometrics.contains(BiometricType.weak)) {
      return 'Biometric';
    }
    return 'Biometric';
  }
}
